### PR TITLE
[clang] Fix crash when declaring invalid lambda member

### PR DESCRIPTION
**Pull request by philnik777**
In valid code, there should only be a very specific set of members in a lambda definition. If the user tries to define something inside the lambda class, this assumption is violated and causes an assertion error. This can be fixed by checking whether the members are valid, and if not, ignore that the class members are potentially unexpected.

I've come across this while working on implementing lambdas in C++03.

### CODE REVIEW

File: clang/lib/AST/DeclCXX.cpp [hunk 1]
```diff
@@ -1526,10 +1526,9 @@ bool CXXRecordDecl::isGenericLambda() const {
 
 #ifndef NDEBUG
 static bool allLookupResultsAreTheSame(const DeclContext::lookup_result &R) {
-  for (auto *D : R)
-    if (!declaresSameEntity(D, R.front()))
-      return false;
-  return true;
+  return llvm::all_of(R, [&](NamedDecl *D) {
+    return D->isInvalidDecl() || declaresSameEntity(D, R.front());
+  });
```
Comment (cor3ntin) [inline clang/lib/AST/DeclCXX.cpp:1531]: Should we move the isInvalidDecl check to declaresSameEntity?
I doubt this is the only place this might happen

**Review by cor3ntin — COMMENTED** [review]
The tests changes look mostly unrelated to this PR

**Comment by shafik** [comment]
Can you add a little more details in the description on the root cause of the bug?

**Comment by philnik777** [comment]
gentle ping

File: clang/lib/AST/DeclCXX.cpp (hunk 1 above)
Comment (shafik) [reply clang/lib/AST/DeclCXX.cpp:1531]: From what I can tell nowhere is checking what they pass to `declaresSameEntity(...)` but I did not dig too deeply to see if we know they are valid earlier in some way.

File: clang/test/SemaCXX/lambda-expressions.cpp [hunk 2]
```diff
@@ -624,22 +626,22 @@ namespace ConversionOperatorDoesNotHaveDeducedReturnType {
...
+      friend U::operator ExpectedTypeU<T>() const; // cxx11-error {{friend declaration of 'operator void (*)(type-parameter-0-0 &)' does not match any declaration}}
 #endif
 
   private:
     int n;
   };
 
   // Should be OK: lambda's call operator is a friend.
-  void use(X &x) { y(x); }
+  void use(X &x) { y(x); } // cxx11-error {{no matching function for call to object}}
```
Comment (AaronBallman) [inline clang/test/SemaCXX/lambda-expressions.cpp:644]: The comment above no longer matches the behavior, the comment should be updated slightly.

File: clang/test/SemaCXX/lambda-expressions.cpp [hunk 3]
```diff
@@ -624,22 +626,22 @@ namespace ConversionOperatorDoesNotHaveDeducedReturnType {
...
+    friend auto T::operator()(int) const; // cxx11-error {{'auto' return without trailing return type; deduced return types are a C++14 extension}}
     friend T::operator ExpectedTypeT() const;
 
     template<typename T>
-      friend void U::operator()(T&) const;
+      friend void U::operator()(T&) const; // cxx11-error {{friend declaration of 'operator()' does not match any declaration}}
     // FIXME: This should not match, as above.
     template<typename T>
-      friend U::operator ExpectedTypeU<T>() const;
+      friend U::operator ExpectedTypeU<T>() const; // cxx11-error {{friend declaration of 'operator void (*)(type-parameter-0-0 &)' does not match any declaration}}
```
Comment (AaronBallman) [inline clang/test/SemaCXX/lambda-expressions.cpp:636]: Oof, that's a bit of a gross diagnostic (not the fault of your patch).

**Review by AaronBallman — COMMENTED** [review]
The changes should come with a release note in clang/docs/ReleaseNotes.rst.

I'm confused -- the patch says that it's fixing a crash, but none of those test cases crash today. Can you add some new test coverage that demonstrates the crash? Also, the changes are to code that's only run in asserts builds... so why are we getting more diagnostics after these changes?

**Review by AaronBallman — CHANGES_REQUESTED** [review]
Oooh, I see now, there's a new RUN line added to the test and that's where the new diagnostics are coming from.

The tests are unrelated to the patch and should be split out into a separate change, and you should add test coverage where we were crashing and we no longer do so. That may need to go into its own file which `// REQUIRES: asserts`.

**Comment by philnik777** [comment]
@AaronBallman Thanks for the review. Sorry for responding so late - I've missed your review.

The tests aren't actually unrelated to the path. This is the file where I discovered the crash, so I just used it to demonstrate that it's fixed. I'm not sure how to split things, since I can't add the run line without fixing the bug.

**Comment by AaronBallman** [comment]
> @AaronBallman Thanks for the review. Sorry for responding so late - I've missed your review.
> 
> The tests aren't actually unrelated to the path. This is the file where I discovered the crash, so I just used it to demonstrate that it's fixed. I'm not sure how to split things, since I can't add the run line without fixing the bug.

Ah, I see now what's going on, and yeah, the tests are not unrelated to the patch (the added RUN line is what causes the crash today), sorry for missing that.

**Review by AaronBallman — APPROVED** [review]
There's a minor nit with a comment in the test file where the comment should be updated, and the changes need a release note. Otherwise, LGTM

**Comment by nico** [comment]
Looks like the included test still makes clang crash: http://45.33.8.238/linux/132722/step_7.txt

**Comment by nico** [comment]
Also on the official waterfall, eg here: https://lab.llvm.org/buildbot/#/builders/188/builds/42935

Can you take a look, and revert for now if it takes a while to fix?

**Pull request merged**